src/pushsource/_impl/backend/pub_source/pub_source.py

```diff
@@ -0,0 +1,106 @@
+import logging
+from concurrent import futures
+
+from more_executors import Executors
+
+from ...helpers import (
+    as_completed_with_timeout_reset,
+    force_https,
+    list_argument,
+    try_int,
+)
+from ...model import AmiPushItem
+from ...source import Source
+from .pub_client import PubClient
+
+LOG = logging.getLogger("pushsource")
+
+
+class PubSource(Source):
+    """Uses push item list from Pub as the source of push items."""
+
+    def __init__(
+        self,
+        url,
+        task_id,
+        threads=4,
+        timeout=60 * 60 * 4,
+    ):
+        """Create a new source.
+
+        Parameters:
+            url (src)
+                Base URL of Pub Tool, e.g. "http://pub.example.com",
+                "https://pub.example.com:8123".
+
+            task_id (int, str, list[str], list[int])
+                Task ID(s) to be used as push item source.
+                If a single string is given, multiple IDs may be
+                comma-separated.
+
+            threads (int)
+                Number of threads used for concurrent queries to Pub.
+
+            timeout (int)
+                Number of seconds after which an error is raised, if no progress is
+                made during queries to Pub.
+        """
+
+        self._url = force_https(url)
+        self._task_ids = self._validate_task_ids(
+            [try_int(x) for x in list_argument(task_id)]
+        )
+        self._client = PubClient(threads=threads, url=self._url)
+        self._timeout = timeout
+        # This executor doesn't use retry because pub-client executor already does that.
+        self._executor = Executors.thread_pool(
+            name="pushsource-pub", max_workers=threads
+        ).with_cancel_on_shutdown()
+
+    def _validate_task_ids(self, task_ids):
+        out = []
+        for task_id in task_ids:
+            if isinstance(task_id, int):
+                out.append(task_id)
+            else:
+                LOG.warning("Invalid task_id: %s", task_id)
+        return out
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._client.shutdown()
+        self._executor.shutdown(True)
+
+    def _push_items_from_json(self, json_obj):
+        out = []
+
+        try:
+            push_items = AmiPushItem._from_data(json_obj)
+            out.extend(list_argument(push_items))
+        except KeyError:
+            LOG.warning("Unsupported or unparsable push item: %s", str(json_obj))
+
+        return out
+
+    def __iter__(self):
+        # Get json Pub responses for all task ids
+        json_fs = [self._client.get_json_f(task_id) for task_id in self._task_ids]
```

Comment:
It looks like all the code around here is going to work fine, but personally I find the naming misleading.

The function names like get_json_f and _push_items_from_json give the impression it's generic and could later be extended for other types, but in reality it's AMI-specific and couldn't be extended to other types cleanly.  For example, if a later content type needs us to start fetching pub-mapfile.json, is get_json_f going to start returning two JSON documents combined and then _push_items_from_json understands how to unpack those?  Would the methods have to be renamed or refactored at that point?

I feel like the way things are split up/named in koji_source is a better approach, which would be something more like `get_ami_json_f` (or maybe `get_json_f("log/images.json")`) followed by `_push_items_from_ami_json`.  Then the natural way to extend it for later types is to add new similar methods which each focus on their supported type.